uvicorn/protocols/websockets/websockets_sansio_impl.py

```diff
@@ -0,0 +1,389 @@
+
+        # WebSocket state
+        self.queue: asyncio.Queue["ASGIReceiveEvent"] = asyncio.Queue()
+        self.handshake_initiated = False
+        self.handshake_complete = False
+        self.close_sent = False
+
+        extensions = []
+        if self.config.ws_per_message_deflate:
+            extensions.append(ServerPerMessageDeflateFactory())
+        self.conn = ServerProtocol(extensions=extensions)
+        self.request: Request
+        self.response: Response
+        self.curr_msg_data_type: str
+
+        self.read_paused = False
+        self.writable = asyncio.Event()
+        self.writable.set()
+
+        # Buffers
+        self.bytes: "bytes" = b""
+
+    def connection_made(self, transport: BaseTransport) -> None:
+        """Called when a connection is made."""
+        transport = typing.cast(Transport, transport)
+        self.connections.add(self)
+        self.transport = transport
+        self.server = get_local_addr(transport)
+        self.client = get_remote_addr(transport)
+        self.scheme = "wss" if is_ssl(transport) else "ws"
+
+        if self.logger.level <= TRACE_LOG_LEVEL:
+            prefix = "%s:%d - " % self.client if self.client else ""
+            self.logger.log(TRACE_LOG_LEVEL, "%sWebSocket connection made", prefix)
+
+    def connection_lost(self, exc: typing.Optional[Exception]) -> None:
+        self.connections.remove(self)
+        if self.logger.level <= TRACE_LOG_LEVEL:
+            prefix = "%s:%d - " % self.client if self.client else ""
+            self.logger.log(TRACE_LOG_LEVEL, "%sWebSocket connection lost", prefix)
+        if self.handshake_initiated and not self.close_sent:
+            self.queue.put_nowait({"type": "websocket.disconnect", "code": 1006})
+
+    def data_received(self, data: bytes) -> None:
+        try:
+            self.conn.receive_data(data)
+        except Exception:
+            self.logger.exception("Exception in ASGI server")
+            self.transport.close()
+        self.handle_events()
+
+    def shutdown(self) -> None:
+        if not self.transport.is_closing():
+            if self.handshake_complete:
+                self.queue.put_nowait({"type": "websocket.disconnect", "code": 1012})
+                self.close_sent = True
+                self.conn.send_close(1012)
+                output = self.conn.data_to_send()
+                self.transport.writelines(output)
+            elif self.handshake_initiated:
+                self.send_500_response()
+                self.queue.put_nowait({"type": "websocket.disconnect", "code": 1006})
+            self.transport.close()
+
+    def handle_events(self) -> None:
+        for event in self.conn.events_received():
+            if isinstance(event, Request):
+                self.handle_connect(event)
+            if isinstance(event, Frame):
+                if event.opcode == websockets.frames.Opcode.CONT:
+                    self.handle_cont(event)
```

Comment:
This line is not covered. as in test suite there is not any test which is sending data in multiple frames. So the whole data is coming in the first frame itself. with fin also set to true. 
Although I had tested it locally on my system.
where I sent multiple frames. in that case, the sequence of frames would be
websockets.frames.Opcode.TEXT or websockets.frames.Opcode.BINARY
followed by ... websockets.frames.Opcode.CONT
may by for websockets.frames.Opcode.CONT frames
and finally in last websockets.frames.Opcode.CONT frame, fin would be set to true

Reply:
Is this related to HTTP/2?

Reply:
@Kludex In case HTTP/1, according to websocket spec, this could be the behaviour

Reply:
got it